### PR TITLE
Refuse to overwrite existing target bundle file unless it was generated by Spago in the first place

### PR DESCRIPTION
**Pull request by fsoikin**
### Description of the change

Fixes #1921

But this is a somewhat breaking change. Not completely breaking, but every existing project will need to either delete the bundle or build it at least once with `--force-overwrite`.

### Checklist:

- [x] Added the change to the "Unreleased" section of the changelog
- ~[ ] Added some example of the new feature to the `README`~
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Spago/Command/Bundle.purs [hunk 1]
```diff
@@ -35,7 +41,7 @@ type RawBundleOptions =
   , extraArgs :: Array String
   }
 
-run :: forall a. Spago (BundleEnv a) Unit
+run :: ∀ a. Spago (BundleEnv a) Unit
```
Comment (fsoikin) [inline src/Spago/Command/Bundle.purs:44]: I reorganized this function a little to reduce noise, I hope it's ok.

File: test-fixtures/bundle-app-node.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env node
-import __module from 'module';import __path from 'path';import __url from 'url';const require = __module.createRequire(import.meta.url);const __dirname = __path.dirname(__url.fileURLToPath(import.meta.url));const __filename=new URL(import.meta.url).pathname
+/* Generated by Spago v0 */import __module from 'module';import __path from 'path';import __url from 'url';const require = __module.createRequire(import.meta.url);const __dirname = __path.dirname(__url.fileURLToPath(import.meta.url));const __filename=new URL(import.meta.url).pathname
```
Comment (fsoikin) [inline test-fixtures/bundle-app-node.js:2]: Unfortunately the comment cannot be on a separate line, because the banner cannot contain newlines, which is in turn because on Windows command-line parameters cannot contain newlines. 

This can be worked around by importing `esbuild` as a Node dependency and calling it directly from PureScript via JS API rather than via command line. But that seemed like too much work for such insignificant gain.

File: bin/src/Flags.purs [hunk 3]
```diff
@@ -69,9 +69,16 @@ outfile =
...
+            <> O.help "Destination path for the bundle. Defaults to ./index.js"
         )
 
+forceOverwrite :: Parser Boolean
+forceOverwrite =
+  O.switch
+    ( O.long "force-overwrite"
+        <> O.help "Overwrite the target bundle file even if it was manually created or generated by a tool other than Spago."
+    )
+
```
Comment (f-f) [inline bin/src/Flags.purs:81]: I think we could shorten this to `--force`, while at the same time having a more specific name to imply the flag is only used for bundling (as other parts of the CLI could use `--force` for other things)
```suggestion
forceBundle :: Parser Boolean
forceBundle =
  O.switch
    ( O.long "force"
        <> O.help "Overwrite the target bundle file even if it was manually created or generated by a tool other than Spago."
    )

```

File: src/Spago/Command/Bundle.purs (hunk 1 above)
Comment (f-f) [reply src/Spago/Command/Bundle.purs:44]: Yes, it looks great!

File: src/Spago/Command/Bundle.purs [hunk 4]
```diff
@@ -47,38 +53,85 @@ run = do
...
+        ]
+      , opts.extraArgs
+      , minify
+      , sourceMap
+      , entrypoint
+      ]
+
+  unless opts.forceOverwrite
+    $ whenM (isNotSpagoGeneratedFile outfile)
+    $ die [ "Target file " <> opts.outfile <> " was not previously generated by Spago. Use --force-overwrite to overwrite anyway." ]
```
Comment (f-f) [inline src/Spago/Command/Bundle.purs:100]: Let's make this change be not breaking in the immediate term by commenting out this check for now, and reintroducing it 6 months down the line. In this way as folks upgrade to the new versions they will start producing bundles with the banner, which will pass the test once we put it back in.

We should leave the tests in too (fixtures and everything), and comment them out too, so they are ready to go when the time comes

```suggestion
  _refuseToOverwriteBundle <- isNotSpagoGeneratedFile outfile
  -- when (refuseToOverwriteBundle && not opts.forceOverwrite) do
  --   die [ "Target file " <> opts.outfile <> " was not previously generated by Spago. Use --force to overwrite anyway." ]
```

File: src/Spago/Command/Bundle.purs [hunk 5]
```diff
@@ -47,38 +53,85 @@ run = do
...
-      , "--outfile=" <> outfile
-      , "--platform=" <> show opts.platform
-      -- See https://github.com/evanw/esbuild/issues/1051
-      , "--loader:.node=file"
-      , format
-      ] <> opts.extraArgs <> minify <> sourceMap <> entrypoint <> nodePatch
+    banner = fold
+      [ spagoGeneratedFilePrefix
+      , " v"
+      , BuildInfo.packages."spago-bin"
```
Comment (f-f) [inline src/Spago/Command/Bundle.purs:79]: I am confused why this version is 0 in the fixtures

Comment (f-f) [reply src/Spago/Command/Bundle.purs:79]: Ah I think it's because the "packages" field will contain the version for the packages being compiled. None of the packages in the fixtures is called "spago-bin", so it comes out as 0.

Since we want to print Spago's version in here, I think we should add a `spagoVersion` field in the `BuildInfo` (I was entirely convinced it was there already). Probably in a separate PR.

File: test-fixtures/bundle-app-node.js (hunk 2 above)
Comment (f-f) [reply test-fixtures/bundle-app-node.js:2]: We can't import esbuild as a node dependency, because we won't be able to bundle it in and distribute that: it's written in Go so it relies on funny JS packaging to be distributed. In any case, this change is fine as is 🙂

File: src/Spago/Command/Bundle.purs (hunk 5 above)
Comment (fsoikin) [reply src/Spago/Command/Bundle.purs:79]: It doesn't "come out" as 0. I manually replaced it with 0.

The problem is, it's supposed to be the "current" version of Spago, which means it can't be any particular version. The current version has to be inserted there during the test run, as that's the only way to get the "current" version. So the "v0" bit is just a placeholder, and it's replaced [here](https://github.com/purescript/spago/pull/1260/files#diff-d4d0ded0a864b48583b9f1fdb5c79023288c9df5fc4bd4499b8d1233b78b5dd9R73-R82), as the comment says.

File: test-fixtures/bundle-app-node.js (hunk 2 above)
Comment (fsoikin) [reply test-fixtures/bundle-app-node.js:2]: Well, an alternative could be to write a JS file that imports and runs esbuild, and then run that file.

File: src/Spago/Command/Bundle.purs (hunk 4 above)
Comment (fsoikin) [reply src/Spago/Command/Bundle.purs:100]: I'm worried that the tests and/or the main code may rot over time from disuse.

How about instead make this check dependent on the presence of a special marker file, and then make the tests write that file so that the check will run during tests.

In addition to the marker file, we could make the check dependent on time (e.g. start running it in 2025) or on current version (e.g. start running after 1.0.0). We'd still have to have the marker file for the tests though.

Comment (fsoikin) [reply src/Spago/Command/Bundle.purs]: I added the marker file logic. Let me know if you would like the time or version condition.

File: bin/src/Flags.purs (hunk 3 above)
Comment (fsoikin) [reply bin/src/Flags.purs]: Done

File: src/Spago/Command/Bundle.purs (hunk 5 above)
Comment (f-f) [reply src/Spago/Command/Bundle.purs:78]: Oh yeah makes sense - thanks for the clarification!

File: src/Spago/Command/Bundle.purs (hunk 4 above)
Comment (f-f) [reply src/Spago/Command/Bundle.purs]: I wouldn't like to create more files. Let's make this dependent on the spago version (+5 patch versions from now or `1.0.0`, whichever comes first) instead. We won't be able to test this now, but we can already check for the watermark, and we can have tests break once we get past the release threshold - I think it's good enough to prevent rot

Comment (f-f) [reply src/Spago/Command/Bundle.purs]: Oh I see - for some reason I entirely missed that comment. Makes sense now!

File: src/Spago/Command/Bundle.purs [hunk 6]
```diff
@@ -47,38 +53,103 @@ run = do
...
+        , "--loader:.node=file" -- See https://github.com/evanw/esbuild/issues/1051
+        , format
+        ]
+      , opts.extraArgs
+      , minify
+      , sourceMap
+      , entrypoint
+      ]
+
+  whenM (FS.exists checkWatermarkMarkerFileName)
```
Comment (f-f) [inline src/Spago/Command/Bundle.purs:97]:
```suggestion
  -- FIXME: remove this after 2024-12-01
  whenM (FS.exists checkWatermarkMarkerFileName)
```

File: test/Spago/Bundle.purs [hunk 7]
```diff
@@ -2,54 +2,94 @@ module Test.Spago.Bundle where
...
+import Spago.Generated.BuildInfo as BuildInfo
 import Test.Spec (Spec)
 import Test.Spec as Spec
+import Test.Spec.Assertions (shouldNotEqual)
+import Test.Spec.Assertions.String (shouldStartWith)
 
 spec :: Spec Unit
 spec = Spec.around withTempDir do
-  Spec.describe "bundle" do
+  Spec.focus $ Spec.describe "bundle" do
```
Comment (f-f) [inline test/Spago/Bundle.purs:16]:
```suggestion
  Spec.describe "bundle" do
```

**Review by f-f — APPROVED** [review]

File: test/Spago/Bundle.purs (hunk 7 above)
Comment (f-f) [reply test/Spago/Bundle.purs]: No worries - I have merged those in quite a few times already, so extra sensitive to them now.

I wonder how we could prevent merging those in at all - maybe spec should allow for focusing using environment variables like many other test runners do, that would remove the need to edit the source at all

Comment (f-f) [reply test/Spago/Bundle.purs]: Yes please! I have been wishing for this in spec for some years already, but my todo list always ends up too long 😄
